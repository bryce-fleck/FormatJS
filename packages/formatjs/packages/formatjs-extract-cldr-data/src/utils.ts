import {Locale} from './types';
import {
  getLocaleHierarchy,
  getAliasesByLang,
  getParentLocalesByLang,
  LDMLPluralRule,
} from '@formatjs/intl-utils';
import {pickBy, isEmpty, isEqual} from 'lodash';

function dedupeUsingParentHierarchy<DataType extends Record<string, any>>(
  ...data: DataType[]
): DataType[] {
  const results: DataType[] = [data.pop()!];
  while (data.length) {
    const datum = data.pop();
    const accumulatedParentData: DataType = Object.assign({}, ...results);
    results.push(
      pickBy<DataType>(
        datum,
        (d, field) => !isEqual(accumulatedParentData[field], d)
      ) as any
    );
  }
  results.reverse();
  return results;
}

export default function generateFieldExtractorFn<
  DataType extends Record<string, any>
>(
  loadFieldsFn: (locale: Locale) => DataType,
  hasDataForLocale: (locale: string) => boolean,
  availableLocales: string[]
) {
  return (locales: Locale[]) => {
    const fieldCache: Record<Locale, DataType> = {};

    // Loads and caches the relative fields for a given `locale` because loading
    // and transforming the data is expensive.
    function populateFields(locale: Locale) {
      if (locale in fieldCache || !hasDataForLocale(locale)) {
        return;
      }

      const lang = locale.split('-')[0];

      const aliases = getAliasesByLang(lang);
      const parentLocales = getParentLocalesByLang(lang);

      const localesIncludingParents = getLocaleHierarchy(
        locale,
        aliases,
        parentLocales
      ).filter(l => availableLocales.includes(l));

      dedupeUsingParentHierarchy(
        ...localesIncludingParents.map(loadFieldsFn)
      ).forEach((datum, i) => {
        fieldCache[localesIncludingParents[i]] = datum;
      });
    }

    locales.forEach(populateFields);
    return pickBy(fieldCache, o => !isEmpty(o));
  };
}

export function collapseSingleValuePluralRule(
  rules: Record<LDMLPluralRule, string>
): string | Record<LDMLPluralRule, string> {
  const keys = Object.keys(rules) as Array<LDMLPluralRule>;
  // dedupe value that looks like `other`
  const uniqueKeys = keys.filter(
    k => k === 'other' || (rules[k] && rules[k] !== rules.other)
  );
  if (uniqueKeys.length === 1) {
    return rules[uniqueKeys[0]];
  }
  return uniqueKeys.reduce((all, k) => {
    all[k] = rules[k];
    return all;
  }, {} as Record<LDMLPluralRule, string>);
}

export const PLURAL_RULES: Array<LDMLPluralRule> = [
  'other',
  'zero',
  'one',
  'two',
  'few',
  'many',
];
