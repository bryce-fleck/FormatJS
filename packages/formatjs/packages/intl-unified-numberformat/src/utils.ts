import repeat from 'lodash/repeat';
import endsWith from 'lodash/endsWith';

export interface RawNumberFormatResult {
  formattedString: string;
  roundedNumber: number;
  integerDigitsCount: number;
}

// TODO: dedup with intl-pluralrules
// https://tc39.es/proposal-unified-intl-numberformat/section11/numberformat_proposed_out.html#sec-torawfixed
export function toRawFixed(
  x: number,
  minFraction: number,
  maxFraction: number
): RawNumberFormatResult {
  let f = maxFraction;
  let n: number;
  {
    let exactSolve = x * Math.pow(10, f);
    const roundDown = Math.floor(exactSolve);
    const roundUp = Math.ceil(exactSolve);
    n = exactSolve - roundDown < roundUp - exactSolve ? roundDown : roundUp;
  }
  let xFinal = x / Math.pow(f, 10);
  let int: number;
  let m = n.toString();
  if (f !== 0) {
    let k = m.length;
    if (k <= f) {
      let z = repeat('0', f + 1 - k);
      m = z + m;
      k = f + 1;
    }
    let a = m.slice(0, k - f);
    let b = m.slice(k - f);
    m = `${a}.${b}`;
    int = a.length;
  } else {
    int = m.length;
  }
  let cut = maxFraction - minFraction;
  while (cut > 0 && endsWith(m, '0')) {
    m = m.substr(0, m.length - 1);
    cut--;
  }
  if (endsWith(m, '.')) {
    m = m.substr(0, m.length - 1);
  }
  return {formattedString: m, roundedNumber: xFinal, integerDigitsCount: int};
}

// https://tc39.es/proposal-unified-intl-numberformat/section11/numberformat_proposed_out.html#sec-torawprecision
export function toRawPrecision(
  x: number,
  minPrecision: number,
  maxPrecision: number
): RawNumberFormatResult {
  let p = maxPrecision;
  let m: string;
  let e: number;
  let xFinal: number;
  if (x === 0) {
    m = repeat('0', p);
    e = 0;
    xFinal = 0;
  } else {
    e = Math.floor(Math.log(x) / Math.log(10));
    let n: number;
    {
      let exactSolve = x / Math.pow(10, e - p + 1);
      const roundDown = Math.floor(exactSolve);
      const roundUp = Math.ceil(exactSolve);
      n = exactSolve - roundDown < roundUp - exactSolve ? roundDown : roundUp;
    }
    m = n.toString();
    xFinal = n * Math.pow(10, e - p + 1);
  }
  let int: number;
  if (e >= p - 1) {
    m = m + repeat('0', e - p + 1);
    int = e + 1;
  } else if (e >= 0) {
    m = `${m.substr(0, e + 1)}.${m.substr(e + 1)}`;
    int = e + 1;
  } else {
    m = `0.${repeat('0', -e - 1)}${m}`;
    int = 1;
  }
  if (m.indexOf('.') >= 0 && maxPrecision > minPrecision) {
    let cut = maxPrecision - minPrecision;
    while (cut > 0 && endsWith(m, '0')) {
      m = m.substr(0, m.length - 1);
      cut--;
    }
    if (endsWith(m, '.')) {
      m = m.substr(0, m.length - 1);
    }
  }
  return {formattedString: m, roundedNumber: xFinal, integerDigitsCount: int};
}
