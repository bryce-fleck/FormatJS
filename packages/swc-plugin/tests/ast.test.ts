import {join} from 'path'
import {compile, FIXTURES_DIR} from './utils'

test('ast', async function () {
  const output = await compile(join(FIXTURES_DIR, `resourcePath.tsx`), {
    ast: true,
    overrideIdFn: (id, defaultMessage, description) => {
      return `HELLO.${id}.${defaultMessage!.length}.${typeof description}`
    },
  })
  expect(output.msgs).toMatchInlineSnapshot(`
    Array [
      Object {
        "defaultMessage": "props {intl}",
        "description": "bar",
        "id": "HELLO..12.string",
      },
      Object {
        "defaultMessage": "this props {intl}",
        "description": "bar",
        "id": "HELLO..17.string",
      },
      Object {
        "defaultMessage": "foo {bar}",
        "description": "bar",
        "id": "HELLO..9.string",
      },
    ]
  `)
  expect(output.code).toMatchSnapshot()
})
