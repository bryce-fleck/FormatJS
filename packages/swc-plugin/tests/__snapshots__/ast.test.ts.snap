// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ast 2`] = `
"function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
export function foo() {
    var _obj;
    props.intl.formatMessage((_obj = {
        id: \\"HELLO..12.string\\",
        defaultMessage: [
            {
                type: 0,
                value: \\"props \\"
            },
            {
                type: 1,
                value: \\"intl\\"
            }, 
        ]
    }, _defineProperty(_obj, \\"defaultMessage\\", \\"props {intl}\\"), _defineProperty(_obj, \\"description\\", \\"bar\\"), _obj), {
        bar: \\"bar\\"
    });
    var _obj1;
    this.props.intl.formatMessage((_obj1 = {
        id: \\"HELLO..17.string\\",
        defaultMessage: [
            {
                type: 0,
                value: \\"this props \\"
            },
            {
                type: 1,
                value: \\"intl\\"
            }, 
        ]
    }, _defineProperty(_obj1, \\"defaultMessage\\", \\"this props {intl}\\"), _defineProperty(_obj1, \\"description\\", \\"bar\\"), _obj1), {
        bar: \\"bar\\"
    });
    var _obj2;
    return intl.formatMessage((_obj2 = {
        id: \\"HELLO..9.string\\",
        defaultMessage: [
            {
                type: 0,
                value: \\"foo \\"
            },
            {
                type: 1,
                value: \\"bar\\"
            }, 
        ]
    }, _defineProperty(_obj2, \\"defaultMessage\\", \\"foo {bar}\\"), _defineProperty(_obj2, \\"description\\", \\"bar\\"), _obj2), {
        bar: \\"bar\\"
    });
}
"
`;
